### DISCUSSION TITLE
Adding UV animation to the StandardMaterial3D would be useful

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#8983] [3D]
At the moment there is no way to animate UVs in the StandardMaterial3D itself.  While you can animate it with an AnimationPlayer, the AnimationPlayer is a node that needs to be added to your scene every time you need to use it and runs on the CPU.  For things like flowing water or lava or other liquids where the UV motion is essentially part of the material, it makes more sense to make this data part of the material rather than the scene.  You could also simply drop your waterfall material onto your waterfall object and then you don't need to set it up any further.

I think the easiest way to do this is to have an optional UV animation section where you can set a vector that is multiplied by the current time and added to the UV1 and UV2 coordinates.  Or perhaps a curve so that users could set up a wave-like animation if they want to the UVs to flow back and forth.

**Comment by Calinou** [comment]
I feel UV animation often needs custom tweaks or layering that a simple "constant scroll" or "sine scroll" options won't suffice. AnimationPlayer also already fits the bill quite well – there's no noticeable overhead to setting a uniform on the CPU every frame. If you really want to skip AnimationPlayer, use a script that sets the value in `_process()`.

Also, it's kind of expected that any water or lava in a full-fledged project will use a custom shader :slightly_smiling_face:
